aws/transport/http/client.go

```diff
@@ -222,3 +215,60 @@ func shallowCopyStruct(src interface{}) interface{} {
 
 	return dstVal.Interface()
 }
+
+func wrapWithoutRedirect(c *http.Client) *http.Client {
+	tr := c.Transport
+	if tr == nil {
+		tr = http.DefaultTransport
+	}
+
+	cc := *c
+	cc.CheckRedirect = limitedRedirect
+	cc.Transport = stubBadHTTPRedirectTransport{
+		tr: tr,
+	}
+
+	return &cc
+}
+
+func limitedRedirect(r *http.Request, via []*http.Request) error {
+	// Request.Response, in CheckRedirect is the response that is triggering
+	// the redirect.
+	resp := r.Response
+	if r.URL.String() == stubBadHTTPRedirectLocation {
+		resp.Header.Del(stubBadHTTPRedirectLocation)
+		return http.ErrUseLastResponse
+	}
+
+	switch resp.StatusCode {
+	case 307, 308:
+		// Only allow 307 and 308 redirects as they preserve the method.
+		return nil
+	}
+
+	return http.ErrUseLastResponse
+}
+
+type stubBadHTTPRedirectTransport struct {
+	tr http.RoundTripper
+}
+
+const stubBadHTTPRedirectLocation = `https://amazonaws.com/badhttpredirectlocation`
+
+func (t stubBadHTTPRedirectTransport) RoundTrip(r *http.Request) (*http.Response, error) {
+	resp, err := t.tr.RoundTrip(r)
+	if err != nil {
+		return resp, err
+	}
+
+	// TODO S3 is the only known service to return 301 without location header.
```

Comment:
S3 would need a separate Buildable Client then?  Or Can this behavior be a BuildableClient option?

Reply:
In general this is block helps prevent the underlying HTTP client returning errors when a `301` HTTP response is received that doesn't contain the location tag. The comment about S3 is background for the reason this even exists, but this logic would be triggered by any 301 API response that doesn't contain a location header.

The Go HTTP client returns a error if an 301 HTTP response is received without a location header, this change prevents that error.